overlord/snapstate/component.go

```diff
@@ -30,8 +31,189 @@ import (
 	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/store"
+	"github.com/snapcore/snapd/strutil"
 )
 
+// InstallComponents installs all of the components in the given names list. The
+// snap represented by info must already be installed, and all of the components
+// in names should not be installed prior to calling this function.
+//
+// TODO:COMPS: respect the transaction that is passed to this function
+func InstallComponents(ctx context.Context, st *state.State, names []string, info *snap.Info, opts Options) ([]*state.TaskSet, error) {
+	var snapst SnapState
+	err := Get(st, info.InstanceName(), &snapst)
+	if err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return nil, &snap.NotInstalledError{Snap: info.InstanceName()}
+		}
+		return nil, err
+	}
+
+	currentComps, err := snapst.CurrentComponentInfos()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, comp := range currentComps {
+		if strutil.ListContains(names, comp.Component.ComponentName) {
+			return nil, fmt.Errorf("cannot install already installed component: %q", comp.Component)
+		}
+	}
+
+	compsups, err := componentSetupsForInstall(ctx, st, names, info, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	snapsup := SnapSetup{
+		Base:        info.Base,
+		SideInfo:    &info.SideInfo,
+		Channel:     info.Channel,
+		Flags:       opts.Flags.ForSnapSetup(),
+		Type:        info.Type(),
+		Version:     info.Version,
+		PlugsOnly:   len(info.Slots) == 0,
+		InstanceKey: info.InstanceKey,
+	}
+
+	setupSecurity := st.NewTask("setup-profiles",
+		fmt.Sprintf(i18n.G("Setup snap %q (%s) security profiles"), info.InstanceName(), info.Revision))
+	setupSecurity.Set("snap-setup", snapsup)
+
+	var kmodSetup *state.Task
+	if requiresKmodSetup(&snapst, compsups) {
+		kmodSetup = st.NewTask("prepare-kernel-modules-components", fmt.Sprintf(
+			i18n.G("Prepare kernel-modules components for %q%s"), info.InstanceName(), info.Revision,
+		))
+	}
+
+	tss := make([]*state.TaskSet, 0, len(compsups))
+	compSetupIDs := make([]string, 0, len(compsups))
+	for _, compsup := range compsups {
+		// since we are installing multiple components, we don't want to setup
+		// the security profiles until the end
+		compsup.MultiComponentInstall = true
+
+		componentTS, err := doInstallComponent(st, &snapst, compsup, snapsup, setupSecurity.ID(), opts.FromChange)
```

Comment:
Maybe it would make sense to pass `setupSecurity` and `kmodSetup` as arguments to `doInstallComponent` and if present, use them instead of creating tasks, I think that would be more future-proof in case tasks change. `MultiComponentInstall` would still be needed though unfortunately.

Reply:
I made this change in 265908c8510ae4a4c5b3f4e4523c6d3d009a14ab, lmk what you think.